pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -308,28 +371,79 @@ func (s *Scheduler) add(ctx context.Context, pod *corev1.Pod) error {
 
 	// Create new node
 	var errs error
+	var compatibleReservedNCT *NodeClaimTemplate
 	for _, nodeClaimTemplate := range s.nodeClaimTemplates {
+		if s.reservedOfferingMode == ReservedOfferingModeStrict && compatibleReservedNCT != nil && compatibleReservedNCT.NodePoolWeight > nodeClaimTemplate.NodePoolWeight {
+			// We tried to schedule the pod to a NodeClaimTemplate previously and it would have scheduled to reserved capacity
+			// had the reservation limit not been exceeded for this scheduling simulation. Since the reserved offering mode is
+			// strict, we shouldn't allow fallback to a NodePool of lesser weight.
+			errs = multierr.Append(errs, NewReservedOfferingError(fmt.Errorf(
+				"skipping evalutation of nodepool %q, pod can schedule to nodepool %q with a greater weight on a reserved offering",
+				nodeClaimTemplate.NodePoolName,
+				compatibleReservedNCT.NodePoolName,
+			)))
+			continue
+		}
+
 		instanceTypes := nodeClaimTemplate.InstanceTypeOptions
 		// if limits have been applied to the nodepool, ensure we filter instance types to avoid violating those limits
 		if remaining, ok := s.remainingResources[nodeClaimTemplate.NodePoolName]; ok {
 			instanceTypes = filterByRemainingResources(instanceTypes, remaining)
 			if len(instanceTypes) == 0 {
-				errs = multierr.Append(errs, fmt.Errorf("all available instance types exceed limits for nodepool: %q", nodeClaimTemplate.NodePoolName))
+				errs = multierr.Append(errs, fmt.Errorf("all available instance types exceed limits for nodepool %q", nodeClaimTemplate.NodePoolName))
 				continue
 			} else if len(nodeClaimTemplate.InstanceTypeOptions) != len(instanceTypes) {
-				log.FromContext(ctx).V(1).WithValues("NodePool", klog.KRef("", nodeClaimTemplate.NodePoolName)).Info(fmt.Sprintf("%d out of %d instance types were excluded because they would breach limits",
-					len(nodeClaimTemplate.InstanceTypeOptions)-len(instanceTypes), len(nodeClaimTemplate.InstanceTypeOptions)))
+				log.FromContext(ctx).V(1).WithValues(
+					"NodePool", klog.KRef("", nodeClaimTemplate.NodePoolName),
+				).Info(fmt.Sprintf(
+					"%d out of %d instance types were excluded because they would breach limits",
+					len(nodeClaimTemplate.InstanceTypeOptions)-len(instanceTypes),
+					len(nodeClaimTemplate.InstanceTypeOptions),
+				))
 			}
 		}
-		nodeClaim := NewNodeClaim(nodeClaimTemplate, s.topology, s.daemonOverhead[nodeClaimTemplate], instanceTypes)
+
+		nodeClaim := NewNodeClaim(nodeClaimTemplate, s.topology, s.daemonOverhead[nodeClaimTemplate], instanceTypes, s.reservationManager, s.reservedOfferingMode)
 		if err := nodeClaim.Add(pod, s.cachedPodData[pod.UID]); err != nil {
-			nodeClaim.Destroy() // Ensure we cleanup any changes that we made while mocking out a NodeClaim
-			errs = multierr.Append(errs, fmt.Errorf("incompatible with nodepool %q, daemonset overhead=%s, %w",
+			nodeClaim.Destroy()
+			if IsReservedOfferingError(err) {
```

Comment:
Rather than worrying about NodeClaims of equal weight and handling that within one scheduling loop, can we just bail-out here for this pod when we get an error like this and we are in non-fallback mode? Doesn't that mean that whichever NodeClaim was ordered first wasn't compatible and that we should try later for the pod in another scheduling loop? I feel like that might also simplify some of your logic here

Reply:
I could go either way on this one. On one hand, attempting to schedule to all equal-weight NodePools will minimize provisioning latency. On the other hand, it may make certain issues more difficult to debug since which NodePool is used will no longer be quite as consistent, though the alphabetical ordering for NodePools of equal weight is an implementation detail and could change anyway. Personally, I'd lean toward the reduced latency making this worth it, but given the reduced complexity and predictability arguments I could be convinced otherwise. What do you think?